src/umode.rs

```diff
@@ -287,35 +297,59 @@ impl UmodeTask {
         Ok(())
     }
 
-    /// Send a request and execute U-mode until an error is returned.
-    pub fn send_req(req: UmodeRequest) -> Result<(), Error> {
+    fn execute_request<T: DataInit>(
+        req: UmodeRequest,
+        shared_data: Option<T>,
+    ) -> Result<(), Error> {
         let mut task = PerCpu::this_cpu().umode_task_mut();
         req.to_registers(task.arch.umode_regs.gprs.a_regs_mut());
+        if let Some(data) = shared_data {
+            PerCpu::this_cpu()
+                .page_table()
+                .share_with_umode(data)
+                .map_err(Error::HypMap)?;
+        }
         let ret = task.run();
         if let Err(e) = &ret {
             match e {
-                Error::UnexpectedTrap | Error::Panic => {
+                ExecError::UnexpectedTrap | ExecError::Panic => {
                     // Reset on unrecoverable u-mode error.
                     // 1. restore memory to original state
                     PerCpu::this_cpu().page_table().restore_umode();
                     // 2. setup umode again.
                     // Panic if we can't restore umode: the hypervisor is in an unrecoverable state.
                     task.reset().expect("Failed to recover U-mode.");
                 }
-                Error::Umode(_) => {}
+                ExecError::Umode(_) => {}
             }
         }
-        ret
+        if shared_data.is_some() {
```

Comment:
Another rust language nerd comment just for consideration, no need to implement: The share_with_umode / clear_umode_shared bracket here would seem to be a good fit for a RAII object - if share_with_umode() were to return an object that clears the umode buffer when it gets dropped, you would have more confidence that the clearing can't accidentally get bypassed.

Reply:
Yes, thoguht about this too. But then I thought that I could get away with defining yet another type that had to be looked up when reading the code by bolting the behavior in send_req, which is I think the right interface.